### DISCUSSION TITLE
Can vcsim be used for manage real ESXi clusters/nodes

### DISCUSSION THREAD

**Discussion by thakkarjigneshX** [vmware/govmomi#2914] [Community Help] [answered]
Hello, I'm a newbie to vcsim world. I did my hands-on for the vcsim to do some basic CRUD operations. However, when I do following, don't see the expected result of VM creation.

> sudo /~/go/bin/vcsim -l <ip>:8989 -username user -password pass &
newpid=`ps -aef | grep vcsim | grep -v sh | grep -v grep | awk '{print $2}'`
export GOVC_URL=https://user:pass@ip:8989/sdk GOVC_SIM_PID=$newpid
export GOVC_INSECURE=1
govc datacenter.create TestDc
govc cluster.create -dc=TestDc TestCluster
govc cluster.add -dc=TestDc -hostname x.x.x.x -username user -password pass -noverify
govc vm.create -dc=TestDc TestVm1

When I create TestVm1 on real ESXi host x.x.x.x, I don't see the VM actually getting created.
Is this expected? We just create simulated entities and not the real one given that hardware is real?

**Comment by embano1** [comment]
Not sure I follow the question. vcsim is a simulator and won't perform any actions in real environments.

**Reply by thakkarjigneshX** [reply]
@embano1 agree. vcsim will not be able to perform any actions for real env. Can govc do this task? I mean can govc connect to real hosts and perform actions on it? Like vm create, delete, power on/off etc?

**Reply by embano1** [reply]
govc can perform certain actions against ESX, though might be limited depending on the APIs you call.

**Reply by thakkarjigneshX** [reply]
I was referring to https://github.com/vmware/govmomi/tree/master/govc to set up govc. 
Do we have any other SDK [say python based] which can perform similar actions against real ESXi?
Also, is there any reference automation available to use govc for performing actions on real ESXi?

**Comment by kbrock** [comment]
Is the question: "Can **govc** be used to manager real ESXi clusters/nodes?
(if so, could you update the title?)

And I think your last statement is: requesting using `govc` to setup `vcsim`.

For terminoligy sake:
- `vcsim` is a vCenter simulator. It pretends to be the server.
- `govc` is a command line interface written in go. It speaks to a server (whether a real vCenter or a simulator)

Now as for sdks google told me there is a lot of documentation out there:
- https://developer.vmware.com/web/sdk/vsphere-automation-python/
- https://github.com/vmware/pyvmomi
- other sdks listed at https://developer.vmware.com/sdks/

**Comment by dougm** [comment, marked as answer]
> When I create TestVm1 on real ESXi host x.x.x.x, I don't see the VM actually getting created

That's right, vcsim does **not** connect to the real ESXi host. Nor is it possible to connect vcsim hosts to a real vCenter.
But both questions have come up a few times over the years, we should at least add some FAQ/doc on this.
Along these lines, @hickeng implemented ESXi host simulation (which also need some docs?), but is container based as we currently [support for VMs](https://github.com/vmware/govmomi/wiki/vcsim-features#containers-as-vms)
